astropy/coordinates/representation.py

```diff
@@ -1593,6 +1618,55 @@ def represent_as(self, other_class, differential_class=None):
 
         return super().represent_as(other_class, differential_class)
 
+    def transform(self, matrix):
+        r"""Transform the unit-spherical coordinates using a 3x3 matrix.
+
+        This returns a new representation and does not modify the original one.
+        Any differentials attached to this representation will also be
+        transformed.
+
+        Parameters
+        ----------
+        matrix : (3,3) array-like
+            A 3x3 (or stack thereof) matrix, such as a rotation matrix.
+
+        Returns
+        -------
+        `UnitSphericalRepresentation` or `SphericalRepresentation`
+            If ``matrix`` is O(3) -- :math:`M \dot M^T = I` -- like a rotation,
+            then the result is a `UnitSphericalRepresentation`.
+            The differentials' classes are also preserved.
+            All other matrices will change the distance, so the dimensional
+            representation is used instead.
+            Likewise, the classes of the differentials will be made dimensional.
+
+        """
+        # the transformation matrix does not need to be a rotation matrix,
+        # so the unit-distance is not guaranteed. For speed, we check if the
+        # matrix is in O(3) and keeps lengths.
+        if np.all(is_O3(matrix)):  # remain in unit-rep
+            if self.differentials:
+                # TODO! shortcut if there are differentials.
+                # Currently just super, which uses Cartesian backend.
+                rep = super().transform(matrix)
+
+            else:
+                xyz = erfa_ufunc.s2c(self.lon, self.lat)
+                p = erfa_ufunc.rxp(matrix, xyz)
+                lon, lat = erfa_ufunc.c2s(p)
+                rep = self.__class__(lon=lon, lat=lat)
+
+        else:  # route through dimensional representation
+            diff_cls = {k: getattr(d, '_dimensional_differential', d.__class__)
+                        for k, d in self.differentials.items()}  # UnitX -> X
+            diffs = self._re_represent_differentials(self, diff_cls)
```

Comment:
I guess my comment was in part that once `SphericalRepresentation` manages to deal with differentials, it will already have to be able to deal with `Unit*Differential` anyway, so I'm not sure it is worth recasting the differentials here.
Overall, it may be easier to just do the `if self.differentials` outside of this `is_O3` check.

Reply:
Do you mean?

```python
        if self.differentials:
            rep = super().transform(matrix)
        elif np.all(is_O3(matrix)):  # remain in unit-rep
            ...
            rep = ...
        else:  # route through dimensional representation
            diff_cls = {k: getattr(d, '_dimensional_differential', d.__class__)
                        for k, d in self.differentials.items()}  # UnitX -> X
            diffs = self._re_represent_differentials(self, diff_cls)
            # ^ works with `self` instead of new_rep b/c compatible diffs
            rep = self._dimensional_representation(
                lon=self.lon, lat=self.lat, distance=1, differentials=diffs
            ).transform(matrix)

        return rep
```

Reply:
Oops, without the differentials in the else block

Reply:
Yes, that's what I had in mind. Just keep it simple here and add special cases of `.transform` for the case of no differentials only.